src/main/cpp/loggingevent.cpp

```diff
@@ -297,40 +302,62 @@ LoggingEvent::KeySet LoggingEvent::getPropertyKeySet() const
 }
 
 
-const LogString LoggingEvent::getCurrentThreadName()
+const LogString& LoggingEvent::getCurrentThreadName()
 {
-#if APR_HAS_THREADS
-	LOG4CXX_THREAD_LOCAL LogString thread_name;
+#if defined(_WIN32)
+	using ThreadIdType = DWORD;
+	ThreadIdType threadId = GetCurrentThreadId();
+#elif APR_HAS_THREADS
+	using ThreadIdType = apr_os_thread_t;
+	ThreadIdType threadId = apr_os_thread_current();
+#else
+	using ThreadIdType = int;
+	ThreadIdType threadId = 0;
+#endif
 
-	if ( thread_name.size() )
+#if defined(__BORLANDC__)
```

Comment:
Do you have access to a Borland compiler for this?  I know that Thorsten uses a Borland-derived compiler.

I ask because I was thinking about this the other day, there is already a ThreadLocal class in the sources.  I was figuring that it was obsolete at this point, but maybe it's still useful for Borland compilers?  Maybe make it a templated class(`ThreadLocal<LogString>`)?

Reply:
I do not have a Borland compiler. My testing approach is to change the "#if __BORLANDC__" to "#if !__BORLANDC__", build and run tests

Reply:
A templated class(ThreadLocal<LogString>) may be worthwhile if there is a second need.

Reply:
Not sure I'm able to follow this discussion: What's the overall reason of introducing the BORLANDC-specific code at all? What are you actually trying to target, the IDE or the 32 Bit legacy compiler or the newer CLANG one or ...? The legacy 32 Bit compiler doesn't even fully support C++11, so I pretty much doubt that the available code works at all. What do you think using `#if !BORLANDC` tell you? :-)

Reply:
Using #if !BORLANDC checks that the code compiles and runs - it tells me nothing regarding the limitations of the legacy compiler support for thread local .

Do yo suggest we remove the BORLANDC tests?

Reply:
> Do you suggest we remove the BORLANDC tests?

Yes. Unless it can be tested and really addresses known limitations of the old legacy compiler, new code targeting that shouldn't be introduced. Changing the condition and testing with some other compiler doesn't mean anything, while the introduced code makes maintenance unnecessary difficult. I wasn't able to successfully build log4cxx using the legacy compiler for a long time already. The following is the docs for supporting the keyword `auto` and it doesn't seem to be supported in BCC32 at all. I very much doubt that the lambda is supported as well.

> The C++98 Definition is supported by previous-generation C++ compilers ([BCC32](https://docwiki.embarcadero.com/RADStudio/Sydney/en/BCC32)).
> The C++11 Definition is supported by [C++11-compatible Clang-enhanced C++ compilers](https://docwiki.embarcadero.com/RADStudio/Sydney/en/Clang-enhanced_C%2B%2B_Compilers).

https://docwiki.embarcadero.com/RADStudio/Sydney/en/Auto

Reply:
It seems that `thread_local` is still not supported even in the latest compiler: https://docwiki.embarcadero.com/RADStudio/Alexandria/en/Modern_C%2B%2B_Language_Features_Compliance_Status

This is rather unfortunate, because that is very useful for things like this.

It seems like we should do the following:

- Remove the Borland-specific code that was added here.  Changing the log event to hold a reference instead of a copy is a good change and should be kept
- Have a discussion on the mailing as to the level of C++ support and the compilers supported.  We should be able to support the Borland compilers without too much trouble as long as we keep away from certain features.

Reply:
For the record, the BORLANDC macro was introduced because the following page [BCC compliance statas](http://docwiki.embarcadero.com/RADStudio/Sydney/en/Modern_C%2B%2B_Language_Features_Compliance_Status) states thread_local is (strangly) only supported by [BCCIOSARM64](https://docwiki.embarcadero.com/RADStudio/Sydney/en/BCCIOSARM64).

Any idea why they would remove support thread_local on some architectures?

Reply:
From my understanding, `__BORLANDC__` either targets the product as whole including the legacy C++ compiler or the legacy compiler only. Your added code is most likely only compatible with the CLANG based compiler, which can be targeted using `__clang__`. So either the condition needs to be made more explicit to target Borland+Clang without ARM or simply only add such compatibility if someone really asks for it and is then able to test it as well.